tempodb/encoding/vparquet2/nested_set_model.go

```diff
@@ -0,0 +1,105 @@
+package vparquet2
+
+import "github.com/grafana/tempo/pkg/util"
+
+func assignNestedSetModelBounds(trace *Trace) {
+	var (
+		assignmentNeeded bool
+		rootSpans        []*wrappedSpan
+		spanChildren     = map[[8]byte][]*Span{}
+	)
+
+	// find root spans and map span IDs to child spans
+	for _, rs := range trace.ResourceSpans {
+		for _, ss := range rs.ScopeSpans {
+			for i, s := range ss.Spans {
+				if s.NestedSetLeft == 0 || s.NestedSetRight == 0 {
+					assignmentNeeded = true
+				}
+
+				if len(s.ParentSpanID) == 0 {
+					rootSpans = append(rootSpans, &wrappedSpan{span: &ss.Spans[i], id: util.SpanIDToArray(s.SpanID)})
+				} else {
+					parentID := util.SpanIDToArray(s.ParentSpanID)
+					spanChildren[parentID] = append(spanChildren[parentID], &ss.Spans[i])
```

Comment:
i wonder if we can do any preallocating or slice reuse here. that may be pre-optimizing and not worth it.

Reply:
I think this would be possible. I'm not so sure about the performance benefits though: It would require another iteration through all spans with a map lookup for every span. In the benchmark above number of allocations did not increase much when calling `assignNestedSetModelBounds`

Reply:
Ignore for now. I'd say this is not worth it unless it starts showing up on profiles.